### DISCUSSION TITLE
Does phase4 prevent sending out invalid messages

### DISCUSSION THREAD

**Discussion by kukel** [phax/phase4#59] [Q&A] [answered]
I was (am?) under the impression that an AP should officially not send out messages that are invalid. I cannot seem to find that in the official Peppol documentation.

I noticed in some (non) test classes with a main like MainPhase4PeppolSenderHelger that they seem to prevent the AP to send out messages that are invalid. Is this according to the rules that are also defined for MLR https://docs.peppol.eu/poacc/upgrade-3/profiles/36-mlr/ Is that a correct assumption? If so, that would be a good USP for Phase4 to explicitly state that it does that. If not, is it possible to extend that?

**Comment by phax** [comment, marked as answer]
Well, phase4 is a multi-purpose AS4 library and not limited to Peppol.
One of the features of the "phase4-peppol-client" though is *to be able to* validate before you send. Software solutions are diverse - as are the requirements ;-)

The rule for senders to validate is in the Peppol Compliance Policy: https://peppol.eu/downloads/compliance/ (chapter 3 if I am recalling it correctly).

**If** you choose to validate with the validation library integrated into phase4 (which is https://github.com/phax/phive and the rules are part of https://github.com/phax/phive-rules), you can validate all the formats that are supported. **But** you of course need to take care about validation rule updates from OpenPeppol side (every 6 month).

The rules for the MLR apply on the receiving side only (as it is the "Message Level Response"). Of course you can also integrate the validation into the receiving side if you like. But in general yes, it would be fairly straight forward to also send an MLR back (but I would recommend NOT to do it directly in the receiving handler, but in a separate thread, because that could lead to the case, that the original sender receives the MLR before it receives the synchronous SignalMessage on the original Invoice).

**Reply by kukel** [reply]
Totally agree with the moment of sending the MLR.  And the mentioning of MLR was more in the context of wondering whether the rules used in the validation before sending are the same as the ones that would be required for MLR.
